src/Matrix.hs

```diff
@@ -106,39 +109,38 @@ fromLayers (x:y:xs) generator = Matrix y x randomVector : fromLayers (y:xs) seco
     (firstGenerator, secondGenerator) = R.split generator
     randomVector = U.generateVector (x * y) firstGenerator
 
-fromList :: Int -> Int -> [a] -> Either String (Matrix a)
+fromList :: DV.Unbox a => Int -> Int -> [a] -> Either String (Matrix a)
 fromList rows columns list
     | (rows == columns && rows == length list) || (rows * columns == length list) = Right $ unsafeFromList rows columns list
     | otherwise = Left "Mismatch between dimensions and list length"
 
-fromVector :: RealFloat a => Int -> Int -> DV.Vector a -> Either String (Matrix a)
+fromVector :: (DV.Unbox a, RealFloat a) => Int -> Int -> DV.Vector a -> Either String (Matrix a)
 fromVector rows columns vector = fromList rows columns $ DV.toList vector
 
-generate :: Int -> Int -> ((Int, Int) -> a) -> Matrix a
+generate :: DV.Unbox a => Int -> Int -> ((Int, Int) -> a) -> Matrix a
 generate rows columns function =
     Matrix rows columns $ DV.generate (rows * columns) (\indice -> function (indice `div` columns, indice `mod` columns))
 
-maximum :: RealFloat a => Matrix a -> a
+maximum :: (DV.Unbox a, RealFloat a) => Matrix a -> a
 maximum (Matrix _ _ vector) = DV.maximum vector
 
-minimum :: RealFloat a => Matrix a -> a
+minimum :: (DV.Unbox a, RealFloat a) => Matrix a -> a
 minimum (Matrix _ _ vector) = DV.minimum vector
 
-normalize :: RealFloat a => Maybe a -> Maybe a -> Matrix a -> Matrix a
-normalize maybeUpperBound maybeLowerBound matrix =
-    if upperBound == lowerBound
-        then const 0 <$> matrix
-        else (\x -> (2 * x - upperBound - lowerBound) / (upperBound - lowerBound)) <$> matrix
-  where
-    upperBound = maybe (maximum matrix) id maybeUpperBound
-    lowerBound = maybe (minimum matrix) id maybeLowerBound
+normalize :: (DV.Unbox a, RealFloat a) => Maybe a -> Maybe a -> Matrix a -> Matrix a
+normalize maybeUpperBound maybeLowerBound matrix = Matrix.map transform matrix
+  where transform  = if upperBound == lowerBound
+                       then const 0
+                       else \x -> (2 * x - upperBound - lowerBound) / (upperBound - lowerBound)
+        upperBound = fromMaybe (maximum matrix) maybeUpperBound
+        lowerBound = fromMaybe (minimum matrix) maybeLowerBound
 
 infixr 8 !
```

Comment:
Should be closer to the begining of the file ?

Reply:
:+1: